backend/src/utils/sso.rs

```diff
@@ -0,0 +1,182 @@
+}
+
+async fn is_valid<'r>(
+    request: &'r Request<'_>,
+    token: &str,
+    con: &mut redis::aio::Connection,
+) -> ValidToken {
+    let redis_result: Result<u32, redis::RedisError> = redis::cmd("EXPIRE")
+        .arg(token)
+        .arg(4 * 3600)
+        .query_async(con)
+        .await;
+    match redis_result {
+        // token exists
+        Ok(1) => {
+            let get_result: Result<i64, redis::RedisError> =
+                redis::cmd("GET").arg(token).query_async(con).await;
+            match get_result {
+                Ok(id) => ValidToken::Valid(id),
+                _ => ValidToken::DatabaseErr,
+            }
+        }
+        // token does not exist
+        Ok(_) => {
+            // hash token to refresh token
+            let mut hash_sha3 = Sha3::sha3_384();
+            hash_sha3.input_str(token);
+            let refresh_token = hash_sha3.result_str();
+            // generate new token
+            let new_token: String = thread_rng()
+                .sample_iter(&Alphanumeric)
+                .take(32)
+                .map(char::from)
+                .collect();
+            // hash new token to new refresh token
+            let mut hash_sha3 = Sha3::sha3_384();
+            hash_sha3.input_str(&new_token);
+            let new_refresh_token = hash_sha3.result_str();
+            // try to rename old refresh token to new refresh token
+            let refresh_result: Result<u32, redis::RedisError> = redis::cmd("RENAMENX")
+                .arg(&refresh_token)
+                .arg(&new_refresh_token)
+                .query_async(con)
+                .await;
+            match refresh_result {
+                // successfully rename refresh token to new refresh token
+                Ok(1) => {
+                    // find id by get refresh_token
+                    let get_result: Result<i64, redis::RedisError> = redis::cmd("GET")
+                        .arg(&new_refresh_token)
+                        .query_async(con)
+                        .await;
+                    let id: i64 = match get_result {
+                        Ok(id) => {
+                            // set id -> new_token
+                            let old_token_get: Result<String, redis::RedisError> =
+                                redis::cmd("GETSET")
+                                    .arg(id)
+                                    .arg(&new_token)
+                                    .query_async(con)
+                                    .await;
+                            // clear old_token -> id
+                            match old_token_get {
+                                Ok(old_token) => {
+                                    let _: Result<i64, redis::RedisError> =
+                                        redis::cmd("DEL").arg(&old_token).query_async(con).await;
+                                }
+                                _ => return ValidToken::DatabaseErr,
+                            };
```

Comment:
Here I think better to use `Option<String>` instead of `String`?

Because if the key does not exist(id->token is missing somehow), the result will be `nil`(none). so if old_token from match of old_token_get is empty(or "" string), it is better not perform a del operation.

You can ask me again if you want to discuss this part with me.

Reply:
I don't know what redis will react when using an empty string as the key of DEL. If it's handled simply as a key that doesn't exist in redis, DEL operation will be ignored in that case. 
I can replace this with Option<String> if necessary though.

Reply:
It is better for you to check this. In sample here are examples of save and get from redis. You can simply change the get key to a empty string to check this and match the result to see how it performs.

Reply:
I have tested using "" as key to set and get redis, and it turns out that "" will be treated just as a common key, so I think there won't be a problem then?

Reply:
Ok. This need no more check. You can resolve this conversation.